File: test_a2a_ws.js
#!/usr/bin/env node
/**
 * WebSocket client to test the A2A interface.
 * Tests JSON-RPC A2A protocol over WebSocket.
 */

const WebSocket = require('ws');

const URI = process.argv[2] || 'ws://127.0.0.1:8080';

async function testA2AWebSocket(uri) {
    console.log('='.repeat(60));
    console.log('A2A WebSocket Interface Test');
    console.log('='.repeat(60));
    console.log(`Target: ${uri}\n`);

    return new Promise((resolve, reject) => {
        console.log(`Connecting to ${uri}...`);
        
        const ws = new WebSocket(uri);

        ws.on('open', () => {
            console.log('✅ Connected to WebSocket server\n');

            // Test 1: Simple A2A request
            console.log('='.repeat(60));
            console.log('Test 1: Simple A2A Request');
            console.log('='.repeat(60));

            const request = {
                jsonrpc: "2.0",
                id: "test-1",
                method: "message/send",
                params: {
                    message: {
                        kind: "message",
                        messageId: "msg-test-1",
                        role: "user",
                        parts: [
                            {
                                kind: "text",
                                text: "Hello World"
                            }
                        ],
                        contextId: null,
                        taskId: null,
                        referenceTaskIds: [],
                        extensions: [],
                        metadata: {
                            method: "greet"
                        }
                    },
                    configuration: null,
                    metadata: {
                        method: "greet"
                    }
                }
            };

            console.log('\n📤 Sending A2A request:');
            console.log(JSON.stringify(request, null, 2));

            ws.send(JSON.stringify(request));
        });

        let testCount = 0;
        const maxTests = 4;
        let pongReceived = false;

        ws.on('message', (data) => {
            testCount++;
            console.log('\n📥 Received response:');
            try {
                const response = JSON.parse(data.toString());
                console.log(JSON.stringify(response, null, 2));

                // Validate response structure
                if (response.jsonrpc) {
                    console.log('\n✅ Valid JSON-RPC response received');
                    if (response.result) {
                        console.log('✅ Response contains result');
                    } else if (response.error) {
                        console.log(`⚠️  Response contains error: ${JSON.stringify(response.error)}`);
                    }
                } else {
                    console.log('❌ Invalid response format');
                }

                // Send next test
                if (testCount === 1) {
                    // Test 2: Invalid JSON-RPC version
                    console.log('\n' + '='.repeat(60));
                    console.log('Test 2: Invalid JSON-RPC version');
                    console.log('='.repeat(60));

                    const invalidRequest = {
                        jsonrpc: "1.0", // Invalid version
                        id: "test-2",
                        method: "message/send",
                        params: {
                    message: {
                        kind: "message",
                        messageId: "msg-test-2",
                                role: "user",
                                parts: [{ kind: "text", text: "test" }],
                        contextId: null,
                        taskId: null,
                        referenceTaskIds: [],
                                extensions: [],
                                metadata: null
                            },
                            configuration: null,
                            metadata: null
                        }
                    };

                    console.log('\n📤 Sending invalid request:');
                    console.log(JSON.stringify(invalidRequest, null, 2));

                    ws.send(JSON.stringify(invalidRequest));
                } else if (testCount === 2) {
                    // Test 3: Malformed JSON
                    console.log('\n' + '='.repeat(60));
                    console.log('Test 3: Malformed JSON');
                    console.log('='.repeat(60));

                    console.log('\n📤 Sending malformed JSON...');
                    ws.send('{ invalid json }');
                } else if (testCount === 3) {
                    // Test 4: Ping/Pong keepalive for long-running chats
                    console.log('\n' + '='.repeat(60));
                    console.log('Test 4: Ping/Pong Keepalive');
                    console.log('='.repeat(60));

                    console.log('\n📤 Sending WebSocket ping...');
                    ws.ping();

                    setTimeout(() => {
                        if (pongReceived) {
                            console.log('✅ Pong received (keepalive ok)');
                        } else {
                            console.log('❌ Pong not received within timeout');
                        }
                        console.log('\n' + '='.repeat(60));
                        console.log('✅ All tests completed');
                        console.log('='.repeat(60));
                        ws.close();
                        resolve(pongReceived);
                    }, 2000);
                }
            } catch (e) {
                console.log(`\n❌ Error parsing response: ${e.message}`);
                if (testCount >= maxTests) {
                    ws.close();
                    resolve(false);
                }
            }
        });

        ws.on('error', (error) => {
            console.error(`\n❌ WebSocket error: ${error.message}`);
            if (error.code === 'ECONNREFUSED') {
                console.error(`\n   Connection refused. Is the server running on ${uri}?`);
                console.error('   Start the server with: ./target/release/baml-rt --api-server');
            }
            reject(error);
        });

        ws.on('close', () => {
            if (testCount < maxTests) {
                console.log('\n⚠️  Connection closed before all tests completed');
            }
        });

        ws.on('pong', () => {
            pongReceived = true;
        });
    });
}

// Main execution
testA2AWebSocket(URI)
    .then((success) => {
        process.exit(success ? 0 : 1);
    })
    .catch((error) => {
        console.error(`\n❌ Test failed: ${error.message}`);
        process.exit(1);
    });

File: test_a2a_ws.py
#!/usr/bin/env python3
"""
WebSocket client to test the A2A interface.
Tests JSON-RPC A2A protocol over WebSocket.
"""

import asyncio
import json
import sys
from typing import Optional

try:
    import websockets
except ImportError:
    print("Error: websockets library not found. Install with: pip install websockets")
    sys.exit(1)


async def test_a2a_websocket(uri: str = "ws://127.0.0.1:8080"):
    """Test A2A WebSocket interface with a simple request."""

    print(f"Connecting to {uri}...")

    try:
        async with websockets.connect(uri) as websocket:
            print("✅ Connected to WebSocket server")

            # Test 1: Simple A2A request with a JavaScript function call
            # This matches the format from the test cases
            request = {
                "jsonrpc": "2.0",
                "id": "test-1",
                "payload": {
                    "method": "message/send",
                    "params": {
                        "message": {
                            "kind": "message",
                            "message_id": "msg-test-1",
                            "role": "user",
                            "parts": [
                                {
                                    "text": "Hello World"
                                }
                            ],
                            "context_id": None,
                            "task_id": None,
                            "reference_task_ids": [],
                            "extensions": [],
                            "metadata": {
                                "method": "greet"
                            }
                        },
                        "configuration": None,
                        "metadata": {
                            "method": "greet"
                        }
                    }
                }
            }

            print("\n📤 Sending A2A request:")
            print(json.dumps(request, indent=2))

            await websocket.send(json.dumps(request))

            # Wait for response
            print("\n⏳ Waiting for response...")
            try:
                response = await asyncio.wait_for(websocket.recv(), timeout=5.0)
                print("\n📥 Received response:")
                response_data = json.loads(response)
                print(json.dumps(response_data, indent=2))

                # Validate response structure
                if "jsonrpc" in response_data:
                    print("\n✅ Valid JSON-RPC response received")
                    if "result" in response_data:
                        print("✅ Response contains result")
                    elif "error" in response_data:
                        print(f"⚠️  Response contains error: {response_data['error']}")
                else:
                    print("❌ Invalid response format")

            except asyncio.TimeoutError:
                print("❌ Timeout waiting for response")
                return False

            # Test 2: Invalid JSON-RPC version
            print("\n" + "="*60)
            print("Test 2: Invalid JSON-RPC version")
            print("="*60)

            invalid_request = {
                "jsonrpc": "1.0",  # Invalid version
                "id": "test-2",
                "payload": {
                    "method": "message/send",
                    "params": {
                        "message": {
                            "kind": "message",
                            "message_id": "msg-test-2",
                            "role": "user",
                            "parts": [{"text": "test"}],
                            "context_id": None,
                            "task_id": None,
                            "reference_task_ids": [],
                            "extensions": [],
                            "metadata": None
                        },
                        "configuration": None,
                        "metadata": None
                    }
                }
            }

            print("\n📤 Sending invalid request:")
            print(json.dumps(invalid_request, indent=2))

            await websocket.send(json.dumps(invalid_request))

            try:
                response = await asyncio.wait_for(websocket.recv(), timeout=5.0)
                print("\n📥 Received response:")
                response_data = json.loads(response)
                print(json.dumps(response_data, indent=2))

                if "error" in response_data:
                    print("✅ Correctly rejected invalid request")
                else:
                    print("⚠️  Expected error response")
            except asyncio.TimeoutError:
                print("❌ Timeout waiting for response")

            # Test 3: Malformed JSON
            print("\n" + "="*60)
            print("Test 3: Malformed JSON")
            print("="*60)

            print("\n📤 Sending malformed JSON...")
            await websocket.send("{ invalid json }")

            try:
                response = await asyncio.wait_for(websocket.recv(), timeout=5.0)
                print("\n📥 Received response:")
                response_data = json.loads(response)
                print(json.dumps(response_data, indent=2))

                if "error" in response_data:
                    error = response_data.get("error", {})
                    if isinstance(error, dict) and error.get("code") == -32700:
                        print("✅ Correctly rejected malformed JSON (Parse error)")
                    else:
                        print(f"⚠️  Got error but unexpected code: {error}")
                else:
                    print("⚠️  Expected error response for malformed JSON")
            except asyncio.TimeoutError:
                print("❌ Timeout waiting for response")

            print("\n" + "="*60)
            print("✅ All tests completed")
            print("="*60)

            return True

    except websockets.exceptions.InvalidURI:
        print(f"❌ Invalid URI: {uri}")
        return False
    except websockets.exceptions.InvalidStatusCode as e:
        print(f"❌ Connection failed with status {e.status_code}")
        return False
    except ConnectionRefusedError:
        print(f"❌ Connection refused. Is the server running on {uri}?")
        print("   Start the server with: ./target/release/baml-rt --api-server")
        return False
    except Exception as e:
        print(f"❌ Error: {e}")
        import traceback
        traceback.print_exc()
        return False


async def main():
    """Main entry point."""
    uri = sys.argv[1] if len(sys.argv) > 1 else "ws://127.0.0.1:8080"

    print("="*60)
    print("A2A WebSocket Interface Test")
    print("="*60)
    print(f"Target: {uri}")
    print()

    success = await test_a2a_websocket(uri)
    sys.exit(0 if success else 1)


if __name__ == "__main__":
    asyncio.run(main())
